controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+					err := r.setGroupReplicationFailure(instance, logger, err)
+					return reconcile.Result{}, err
+				}
+			}
+		} else {
+			logger.Info("No selector provided for PVCs in volumeGroupReplication source")
+			return reconcile.Result{}, nil
+		}
+
+		// Update PersistentVolumeClaimsRefList in VGR Status
+		tmpRefList := []corev1.LocalObjectReference{}
+		for i, pvc := range pvcList.Items {
+			if i < 100 {
+				tmpRefList = append(tmpRefList, corev1.LocalObjectReference{
+					Name: pvc.Name,
+				})
+			} else {
+				logger.Info("More than 100 PVCs match the given selector, using only first 100 PVCs")
+				break
+			}
+		}
+
+		// Annotate each PVC with owner and add finalizer to it
+		for _, pvc := range pvcList.Items {
+			err = AnnotatePVCWithOwner(r.Client, logger, instance.Name, &pvc, replicationv1alpha1.VolumeGroupReplicationNameAnnotation)
+			if err != nil {
+				logger.Error(err, "Failed to add VGR owner annotation on PVC")
+				return ctrl.Result{}, err
+			}
+
+			if err = AddFinalizerToPVC(r.Client, logger, &pvc, vgrReplicationFinalizer); err != nil {
+				logger.Error(err, "Failed to add VGR finalizer on PersistentVolumeClaim")
+				return reconcile.Result{}, err
+			}
+		}
+
+		if !reflect.DeepEqual(instance.Status.PersistentVolumeClaimsRefList, tmpRefList) {
+			instance.Status.PersistentVolumeClaimsRefList = tmpRefList
+			err = r.Client.Status().Update(ctx, instance)
+			if err != nil {
+				logger.Error(err, "failed to update VolumeGroupReplication resource")
+				err := r.setGroupReplicationFailure(instance, logger, err)
+				return reconcile.Result{}, err
+			}
 
-	return ctrl.Result{}, nil
+			// Get the PV handles for all the PVCs which are provisioned using the same provisioner
+			// as mentioned in the VGRClass resource
+			pvHandlesList, err := r.getPVHandles(*vgrClassObj, pvcList, logger)
+			if err != nil {
+				logger.Error(err, "failed to get volume handles from the pvs")
+				err := r.setGroupReplicationFailure(instance, logger, err)
+				return reconcile.Result{}, err
+			}
+
+			// Create/Fetch VolumeGroupReplicationContent CR
+			if instance.Spec.VolumeGroupReplicationContentName != "" {
+				err = r.Client.Get(ctx, types.NamespacedName{Name: instance.Spec.VolumeGroupReplicationContentName}, vgrContentObj)
+				if err != nil {
+					logger.Error(err, "failed to get volumeGroupReplicationContent resource", "VGRContentName", instance.Spec.VolumeGroupReplicationContentName)
+					err := r.setGroupReplicationFailure(instance, logger, err)
+					return reconcile.Result{}, err
+				}
+				if !reflect.DeepEqual(pvHandlesList, vgrContentObj.Spec.Source.VolumeHandles) {
+					vgrContentObj.Spec.Source.VolumeHandles = pvHandlesList
+					err = r.Client.Update(ctx, vgrContentObj)
+					if err != nil {
+						logger.Error(err, "failed to update volumeGroupReplicationContent resource", "VGRContentName", vgrContentObj.Name)
+						err := r.setGroupReplicationFailure(instance, logger, err)
```

Comment:
err will be nil here if `setGroupReplicationFailure` is successful, return actual error if update fails, revisit this in all places